network/topics.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+// Topic is a key-value pair
+type Topic struct {
```

Comment:
I'd rename Topics to something else to not confuse with MQ topics...
In fact it is used as a dict (** GetValue** the only method). Although I probably do not understand all the future use cases but it looks like map is more suitable then an array of struct unless we expect duplicate keys.

Reply:
I was thinking about map for Topics too; the issue is that we want to make sure that the encoding of the Topics would consistently generate the same hash every time. This isn't guaranteed with a map, where the order is arbitrary.

Reply:
What do we expect in topics? if there are many keys then GetValue might be an issue. In this case (and if the keys are unique) an ordered map might work:
```
type OrderedMap struct {
  data map[string][]byte
  keys []string
}

func (om OrderedMap) Put(k string, v []byte) {
  om.keys = append(om.keys, k)
  om.data[k] = v
}

func (om OrderedMap) Serialize() (res []byte) {
  for _, k := range om.keys {
    res = append(res, []byte(k))
    res = append(res, om.data[k])
  }
}
```
But not sure if we buy much here. I'm fine with arrays for small amount of keys.

Reply:
The number of topics is limited to 32. So there will be no more than 32 keys. Doing anything fancier that an array will just overcomplicate things. 

As for duplication, nothing prevents the construction of multiple topics with the same key, and it will be marshalled/unmarshalled properly, but GetValue returns a single value, which will be the first topic with the key value.

Reply:
I expect to have around 3 keys.